src/v/redpanda/application.cc

```diff
@@ -354,41 +366,56 @@ void application::shutdown() {
      * possible.
      */
     if (_datalake_coordinator_mgr.local_is_initialized()) {
-        _datalake_coordinator_mgr
-          .invoke_on_all(&datalake::coordinator::coordinator_manager::shutdown)
-          .get();
+        shutdown_with_watchdog(
+          "datalake::coordinator_manager", "shutdown", [this] {
+              return _datalake_coordinator_mgr.invoke_on_all(
+                &datalake::coordinator::coordinator_manager::shutdown);
+          });
     }
     if (_datalake_manager.local_is_initialized()) {
-        _datalake_manager.invoke_on_all(&datalake::datalake_manager::shutdown)
-          .get();
+        shutdown_with_watchdog("datalake_manager", "shutdown", [this] {
+            return _datalake_manager.invoke_on_all(
+              &datalake::datalake_manager::shutdown);
+        });
     }
     if (_datalake_credential_mgr.local_is_initialized()) {
-        _datalake_credential_mgr
-          .invoke_on_all(&datalake::credential_manager::stop)
-          .get();
+        shutdown_with_watchdog("datalake::credential_manager", "stop", [this] {
+            return _datalake_credential_mgr.invoke_on_all(
+              &datalake::credential_manager::stop);
+        });
     }
 
     // Stop all partitions before destructing the subsystems (transaction
     // coordinator, etc). This interrupts ongoing replication requests,
     // allowing higher level state machines to shutdown cleanly.
     if (partition_manager.local_is_initialized()) {
-        partition_manager
-          .invoke_on_all(&cluster::partition_manager::stop_partitions)
-          .get();
+        shutdown_with_watchdog("partition_manager", "stop_partitions", [this] {
+            return partition_manager.invoke_on_all(
+              &cluster::partition_manager::stop_partitions);
+        });
     }
 
     // Wait for all requests to finish before destructing services that may be
     // used by pending requests.
     if (_kafka_server.ref().local_is_initialized()) {
-        _kafka_server.wait_for_shutdown().get();
-        _kafka_server.stop().get();
+        shutdown_with_watchdog("kafka_server", "wait_for_shutdown", [this] {
+            return _kafka_server.wait_for_shutdown();
+        });
+        shutdown_with_watchdog(
+          "kafka_server", "stop", [this] { return _kafka_server.stop(); });
```

Comment:
nit: do we need two arguments? probably just call "kafka_server::stop" or something?

Reply:
>nit: do we need two arguments? probably just call "kafka_server::stop" or something?

yeh. pretty easy to make a copy/paste mistake and then result in a useless log message.

if you do something like `shutdown(_kafka_server, [this](auto& s) { return s.stop(); }));

then in the definiton of shutdown you can do something like derive the name from

```
template<typename T>
shutdown(T&) {
auto name = ss::pretty_type_name(typeid(T));
}
```

and then you can probably even add the line number too either by making it a macro or using std::source_location.